src/evaluator/functions.rs

```diff
@@ -1449,6 +1449,60 @@ pub fn fn_round<'a>(
     Ok(Value::number(context.arena, num))
 }
 
+pub fn fn_reduce<'a>(
+    context: FunctionContext<'a, '_>,
+    args: &[&'a Value<'a>],
+) -> Result<&'a Value<'a>> {
+    max_args!(context, args, 3);
+
+    if args.len() < 2 {
+        return Err(Error::T0410ArgumentNotValid(0, 2, context.name.to_string()));
+    }
+
+    let first_arg = args[0];
+    let func = args[1];
+    let init = args.get(2).copied();
+
+    if !first_arg.is_array() {
+        if first_arg.is_string() {
+            return Ok(first_arg);
+        }
+        return Ok(first_arg);
+    }
+
+    let (elements, _extra_field) = match first_arg {
+        Value::Array(elems, extra) => (elems, extra),
+        _ => return Err(Error::T0410ArgumentNotValid(0, 1, context.name.to_string())),
+    };
+
+    if elements.is_empty() {
+        return Ok(init.unwrap_or_else(|| Value::undefined()));
+    }
+
+    if !func.is_function() {
+        return Err(Error::T0410ArgumentNotValid(1, 1, context.name.to_string()));
+    }
+
+    let mut accumulator = init.unwrap_or_else(|| elements[0]);
+
+    let start_index = if init.is_some() { 0 } else { 1 };
+
+    for (index, value) in elements.iter().enumerate().skip(start_index) {
+        let index_value = Value::number(context.arena, index as f64);
+
+        if accumulator.is_string() && value.is_string() {
+            // Proceed with string concatenation logic if both are strings
+        } else if !accumulator.is_number() || !value.is_number() {
+            return Err(Error::T0410ArgumentNotValid(1, 1, context.name.to_string()));
+        }
+
+        let args_to_pass = vec![accumulator, value, index_value, first_arg];
```

Comment:
We don't need to create a `vec!`. We can create a slice inline.

```rust
accumulator = context.evaluate_function(func, &[accumulator, value, index_value, first_arg]);
```

Reply:
It might also be worth renaming `first_arg`. It is confusing what was being passed in. It's the array. Maybe we could rename it to something like `original_array.`